src/render_output_ppm_bonus.c

```diff
@@ -34,47 +23,56 @@ static void	ft_write_ppm_header(int width, int height, int fd)
 	ft_putchar_fd('\n', fd);
 }
 
-static void	ft_write_to_file(int fd, int *pix_arr, int pixel_sum)
-{
-	int	i;
-
-	i = 0;
-	while (i < pixel_sum)
-	{
-		ft_write_pixel(pix_arr[i++], fd);
-		if (i % 100 == 0)
-			printf("\r%d%%", (i * 100) / pixel_sum);
-	}
-	printf("\r100%%\nFinished printing\n");
-}
-
 static void	ft_create_filename(char outfile[14])
 {
 	static int	i;
 	char		num[3];
 
 	if (i++ > 99)
 		i = 0;
-	ft_strlcpy(outfile, "outfile_", 8);
+	ft_strlcpy(outfile, "outfile_", 9);
 	ft_itoa_in_place(i, num);
 	ft_strlcat(outfile, num, 14);
 	ft_strlcat(outfile, ".ppm", 14);
 }
 
-t_err	ft_output_as_ppm(const t_img img, bool *is_printing)
+static void	ft_write_to_buffer(const t_img img, char *file_buffer)
+{
+	int			i;
+	uint32_t	*pixel;
+
+	pixel = (uint32_t *)img.addr;
+	i = -1;
+	while (i++ < img.size.x * img.size.y - 1)
+	{
+		file_buffer[i * 3] = (*pixel & 0xFF0000) >> 16;
+		file_buffer[i * 3 + 1] = (*pixel & 0x00FF00) >> 8;
+		file_buffer[i * 3 + 2] = (*pixel & 0x0000FF);
+		pixel += 1;
+	}
+}
+
+t_err	ft_output_as_ppm(const t_img img)
 {
 	int		fd;
 	char	outfile[14];
+	char	*file_buffer;
```

Comment:
Speeds up printing immensely.
Instead of writing each pixel value line by line to the file, we write the r,g,b values at first to this (local) buffer. After it has been written, the buffer can be written to file in one go.

Reply:
So the writing process is itself time consuming

Reply:
Yes!
Writing to file is way slower than malloc - "write" to buffer - write to file.
Writing to file needs the syscall write for every line.
Buffer only needs one write() call as "writing" to buffer is not exactly writing but just setting the char array.